runtime/lua/vim/shared.lua

```diff
@@ -4,32 +4,32 @@
 -- test-suite. If, in the future, Nvim itself is used to run the test-suite
 -- instead of "vanilla Lua", these functions could move to src/nvim/lua/vim.lua
 
+local module = {}
 
 --- Returns a deep copy of the given object. Non-table objects are copied as
 --- in a typical Lua assignment, whereas table objects are copied recursively.
 ---
 --@param orig Table to copy
 --@returns New table of copied keys and (nested) values.
-local function deepcopy(orig)
-  error(orig)
-end
-local function _id(v)
-  return v
-end
-local deepcopy_funcs = {
-  table = function(orig)
-    local copy = {}
-    for k, v in pairs(orig) do
-      copy[deepcopy(k)] = deepcopy(v)
-    end
-    return copy
-  end,
-  number = _id,
-  string = _id,
-  ['nil'] = _id,
-  boolean = _id,
-}
-deepcopy = function(orig)
+module.deepcopy = function(orig)
+  local function _id(v)
+    return v
+  end
+
+  local deepcopy_funcs = {
+    table = function(_orig)
+      local copy = {}
+      for k, v in pairs(_orig) do
+        copy[module.deepcopy(k)] = module.deepcopy(v)
+      end
+      return copy
+    end,
+    number = _id,
+    string = _id,
+    ['nil'] = _id,
+    boolean = _id,
+  }
+
   return deepcopy_funcs[type(orig)](orig)
 end
```

Comment:
could use an extra closure, to avoid re-defining `_id` and `deepcopy_funcs` at each call:

```lua
module.deepcopy = function()
  local function _id(v)
    return v
  end

  local deepcopy_funcs = {
    table = function(_orig)
      local copy = {}
      for k, v in pairs(_orig) do
        copy[module.deepcopy(k)] = module.deepcopy(v)
      end
      return copy
    end,
    number = _id,
    string = _id,
    ['nil'] = _id,
    boolean = _id,
  }

  return function(orig)
    return deepcopy_funcs[type(orig)](orig)
  end
end()
```

Reply:
This has syntax error. Lua may not use () for immediate execution after end.

Reply:
just wrap it in parens. see `paste` for example.

Reply:
I see :)